File: worldgen.py
from noise import pnoise2, pnoise3
import random

def generate_world():
    world = {}
    terrain_size = 32
    max_height = 12
    seed = random.randint(0, 1000000)

    def get_biome(x, z):
        biome_value = pnoise2(x * 0.02, z * 0.02, base=seed)
        if biome_value > 0.4:
            return 'mountain'
        if biome_value > -0.1:
            return 'forest'
        return 'plains'

    # Generate terrain
    for x in range(terrain_size):
        for z in range(terrain_size):
            height = int(pnoise2(x * 0.08, z * 0.08, base=seed) * max_height / 2 + max_height / 2)
            biome = get_biome(x, z)

            for y in range(height + 1):
                block_type = 'dirt'
                if y == height:
                    block_type = 'grass'
                elif y < height - 1:
                    block_type = 'coal' if random.random() < 0.15 else 'stone'
                world[f'{x},{y},{z}'] = block_type

            if biome == 'plains' and pnoise2(x * 0.05, z * 0.05, base=seed + 1) < -0.5 and height < 6:
                world[f'{x},{height + 1},{z}'] = 'water'

    # Generate caves
    for x in range(terrain_size):
        for z in range(terrain_size):
            for y in range(max_height):
                if pnoise3(x * 0.15, y * 0.15, z * 0.15, base=seed + 2) > 0.75:
                    key = f'{x},{y},{z}'
                    if key in world:
                        del world[key]

    # Generate trees
    for x in range(1, terrain_size - 1):
        for z in range(1, terrain_size - 1):
            biome = get_biome(x, z)
            if (biome == 'forest' and random.random() < 0.08) or (biome == 'plains' and random.random() < 0.02):
                surface_y = max(
                    int(key.split(',')[1]) for key in world
                    if key.startswith(f'{x},') and key.endswith(f',{z}')
                ) + 1

                trunk_height = random.randint(3, 4)
                for y in range(surface_y, surface_y + trunk_height):
                    world[f'{x},{y},{z}'] = 'dirt'

                foliage_y = surface_y + trunk_height - 1
                for dy in range(2):
                    offset = 1 if dy == 0 else 0
                    for dx in range(-offset, offset + 1):
                        for dz in range(-offset, offset + 1):
                            if random.random() < 0.8:
                                world[f'{x + dx},{foliage_y + dy},{z + dz}'] = 'grass'

    return world

File: worldgen.js
const { createNoise2D, createNoise3D } = require('simplex-noise');

function generateWorld() {
    const world = {};
    const terrainSize = 32;
    const noise2D = createNoise2D();
    const noise3D = createNoise3D();
    const maxHeight = 12;

    // Biome noise
    function getBiome(x, z) {
        const biomeValue = noise2D(x * 0.02, z * 0.02);
        if (biomeValue > 0.4) return 'mountain';
        if (biomeValue > -0.1) return 'forest';
        return 'plains';
    }

    // Generate terrain
    for (let x = 0; x < terrainSize; x++) {
        for (let z = 0; z < terrainSize; z++) {
            const height = Math.floor(noise2D(x * 0.08, z * 0.08) * maxHeight / 2 + maxHeight / 2);
            const biome = getBiome(x, z);

            for (let y = 0; y <= height; y++) {
                let blockType = 'dirt';
                if (y === height) blockType = 'grass';
                else if (y < height - 1) blockType = Math.random() < 0.15 ? 'coal' : 'stone';
                world[`${x},${y},${z}`] = blockType;
            }

            // Add water for rivers
            if (biome === 'plains' && noise2D(x * 0.05, z * 0.05) < -0.5 && height < 6) {
                world[`${x},${height + 1},${z}`] = 'water';
            }
        }
    }

    // Generate caves
    for (let x = 0; x < terrainSize; x++) {
        for (let z = 0; z < terrainSize; z++) {
            for (let y = 0; y < maxHeight; y++) {
                if (noise3D(x * 0.15, y * 0.15, z * 0.15) > 0.75) {
                    delete world[`${x},${y},${z}`];
                }
            }
        }
    }

    // Generate trees
    for (let x = 1; x < terrainSize - 1; x++) {
        for (let z = 1; z < terrainSize - 1; z++) {
            const biome = getBiome(x, z);
            if ((biome === 'forest' && Math.random() < 0.08) || (biome === 'plains' && Math.random() < 0.02)) {
                const surfaceY = Math.max(...Object.keys(world)
                    .filter(key => key.startsWith(`${x},`) && key.endsWith(`,${z}`))
                    .map(key => parseInt(key.split(',')[1]))) + 1;

                // Tree trunk
                const trunkHeight = Math.floor(Math.random() * 2) + 3;
                for (let y = surfaceY; y < surfaceY + trunkHeight; y++) {
                    world[`${x},${y},${z}`] = 'dirt';
                }

                // Foliage
                const foliageY = surfaceY + trunkHeight - 1;
                for (let dy = 0; dy < 2; dy++) {
                    const offset = dy === 0 ? 1 : 0;
                    for (let dx = -offset; dx <= offset; dx++) {
                        for (let dz = -offset; dz <= offset; dz++) {
                            if (Math.random() < 0.8) {
                                world[`${x + dx},${foliageY + dy},${z + dz}`] = 'grass';
                            }
                        }
                    }
                }
            }
        }
    }

    return world;
}

module.exports = { generateWorld };